### ISSUE TITLE
[Question][Data Models] Code Quality / Testing

### ISSUE THREAD

**Issue by martynvdijke** [apache/devlake#7765]
## Question

Hi all,

I have a question I hope you can help me with.
I see from the roadmap over https://devlake.apache.org/docs/Overview/Roadmap/ that Code Quality / Testing is on the roadmap.
Could you maybe explain these data models a bit more and how you plan on getting / ingesting this data?
And your vision and idea behind that is, what you want to measure and how ?

Background:
I want to see if it is possible to measure code coverage and quality metrics next to build time and correlate them.
We have some long ci build pipelines and are curious if we can fine tune the number of tests vs times, or at least gain insights in them at best and maybe split them up or so. 
We have the feeling this can be better in balance at least.

I initially thought of using the webhook for this, but I now found out that the webhooks really are hooked into data models (which makes sense). 
I am not sure if you are thinking about adding a generic webook data model, so that special use cases can or extra data can be added ?

So long story short, can you maybe explain the vision and idea behind the Code Quality / Testing data model ?

Also awsome work on this project so far !

**Comment by d4x1** [comment]
@martynvdijke  Thanks for your issue.

How to process Code Quality is  a difficult question. 

I cannot answer your question at once, but here are some information and my ideas I can provide.

We have only one datasource SonarQube, that belongs to `DOMAIN_TYPE_CODE_QUALITY` at present, and the domain layer tables which as be found https://github.com/apache/incubator-devlake/tree/main/backend/core/models/domainlayer/codequality are mapped from SonarQube's data models directly.

As for your questions,
1. "I am not sure if you are thinking about adding a generic webook data model, so that special use cases can or extra data can be added ?"
You can describe it in detail, adding a webhook is easy, but what do you mean by `adding a GENRIC webhook data model`?
2.  "Fine tune the number of tests vs times"
You can 
    a. try to build a relation: incident -> line of code  -> test, tests with a high incident rate should be run as much as possible. 
    b. calculate the overlap of tests, and mark some tests are more efficient. 

We haven't received enough feedbacks about code quality, so you can give us some advice on it. We can explore it together.

**Comment by martynvdijke** [comment]
Thanks for the response this clears parts up, thanks for taking the time 👍 

1. My tought is that a generic data model / data source could be a nice addition to allow for some customization or realted uses cases that are currently not fully covered.  A use case now would be if we can use a generic webhook data model to upload our python code coverage reports. So that we can use that in a custom dashboard.
2. That sounds like a nice plan, will do some tinkering around to see whats possible

We are also searching for a bit here, but I think high over we would want to at least start with:
1. Upload code coverage metrics of our unit and feature tests (in python)
2. Upload the code complexity score

So that we can track that over time and see our improvments in this area.

Hope this makes a bit more clear ?

**Comment by d4x1** [comment]
@martynvdijke 
Webhook is suitable for posting some additional data.
Let's talk about this `generic data model` first,  it can be added with little effort.
If you post a json data, what is the expected effect?  A new table is created  and all json fields are flattened and have a corresponding column? Or just set a new table like:
| id    | data |
| -------- | ------- |
| x  | {json_data}    |

**Comment by martynvdijke** [comment]
Hey,

I think option 1 would be the nicest it would allow you to query the data in grafana in the best way I think.
Or at least from a user perspective in the best way.

**Comment by d4x1** [comment]
@martynvdijke  Are you interested in submiting a PR?  It doesn't need too much work.

**Comment by martynvdijke** [comment]
Hey,

Yeah I should have some time next week to dive into this and submit a PR for this.

If you have time some pointers on how to get the database formatting from the json would be nice, never done anything like that.

**Comment by d4x1** [comment]
@martynvdijke 

1. Start a new api handler in plugin webhook
2. In this handler
    1. get request body
    2. parse request body
    3. add new tables or new fields in a certain table according to the request data (I think this will help: https://github.com/apache/incubator-devlake/blob/main/backend/plugins/customize/service/service.go#L113)

**Comment by martynvdijke** [comment]
HI @d4x1, sorry it took a bit longer then expected.

I tried to follow your guidelines but I cannot really make it work.
Or cannot really understand the entire flow of how it should fit together and program this in go.

Would it be possible for you to pickup ?

**Comment by martynvdijke** [comment]
OKwey I have some code on https://github.com/martynvdijke/devlake/tree/feat-generic-webhook

**Comment by d4x1** [comment]
@martynvdijke Let me see if  I can help.
